docs/lofi-hifi-backends.md

```diff
@@ -0,0 +1,279 @@
+
+This document focuses primarily on training, specifically `ilab train peft` and `ilab train phased`.
+
+## Key Component
+
+### Building off of the InstructLab Structure Redesign
+
+After github.com/instructlab/instructlab/pull/990 is merged, ilab will use a parent -> child command structure. This proposal operates under that new structure.
+
+This is the proposed new structure under the above Enhancement:
+
+```console
+ilab
+|
+|_______model
+|       |
+|       |____convert
+|       |____download
+|       |____train (--convert)
+|       |____serve   (-i)
+|       |____chat
+|       |____inference
+|       |
+|_______data
+|       |
+|       |____generate
+|       |
+|_______config
+|       |
+|       |____init
+|       |
+|_______tax
+|       |
+|       |____diff
+|       |____check
+|       |____download
+```
+
+And this would be the structure after these new commands are added:
+
+```console
+ilab
+|
+|_______model
+|       |
+|       |____convert
+|       |____download
+|       |____train (--convert)
+|       |    |
+|       |    |______peft *
+|       |    |______phased *
+|       |
+|       |____serve   (-i)
+|       |____chat
+|       |____inference
+|       |
+|_______data
+|       |____generate
+|       |    |
+|       |    |______lofi * (name pending)
+|       |    |______hifi * (name pending)
+|       |
+|       |____mix *
+|       |
+|       |
+|_______checkpoint
+|       |
+|       |____evaluate *
+|       |
+|_______config
+|       |
+|       |____init
+|_______tax
+|       |
+|       |____diff
+|       |____check
+|       |____download
+```
+
+
+The starred commands are the new ones under the redesigned structure. We are opting to add peft and phased as commands under train and lofi/hifi under generate rather than flags given their completely different functionality and backend structure.
+
+`ilab train peft` would utilize much of the existing upstream training code with the following differences:
+
+- deepseed support
+- fsdp support (eventually)
+- full evaluation at the end of training
+- inter-epoch evaluation using pytorch's built in evaluators that fit for value loss primarily.
+
+`ilab train phased` would utlize much of the backend training code. Primarily doing the following:
+
+- running a single phase of training
+- outputting checkpoint dirs for evaluation to be run on
+- deepspeed support
+- fsdp support.
+
+`ilab train phased` will be used in conjunction with `ilab checkpoint evaluate`. This command will run full scale inter-checkpoint evaluation on the given directory. An output dir will then hold then best checkpoint and all necessary data to run another `ilab train phased` command on.
+
+Much of the code for `ilab train phased` and `ilab train peft` will live in the CLI, and hopefully will be shared. However, plans to move much of this code to an external training library should eventually happen. This is most likely not in scope for the first version of this work.
```

Comment:
Is it still the case? Looks like the training repo is around the corner  https://github.com/instructlab/dev-docs/pull/85